src/parser/toplevel.lisp

```diff
@@ -1488,19 +1540,35 @@ consume all attributes")))
                  (stringp (cst:raw (cst:third form))))
         (setf docstring (cst:raw (cst:third form))))
 
-      (make-toplevel-define-instance
-       :context context
-       :pred (parse-predicate unparsed-predicate
-                              (source:make-location source
-                                                    (util:cst-source-range unparsed-predicate)))
-       :docstring docstring
-       :methods (loop :for methods := (cst:nthrest (if docstring 3 2) form) :then (cst:rest methods)
-                      :while (cst:consp methods)
-                      :for method := (cst:first methods)
-                      :collect (parse-instance-method-definition method (cst:second form) source))
-       :location (form-location source form)
-       :head-location (form-location source (cst:second form))
-       :compiler-generated nil))))
+      (let ((methods (loop :with forms := (cst:nthrest (if docstring 3 2) form)
+                           :while (cst:raw forms)
+                           :for inline-p := (let ((form (cst:first forms)))
+                                              (when (and (cst:consp form)
+                                                         (eq 'coalton:inline (cst:raw (cst:first form))))
+                                                (if (null (cst:raw (cst:rest form)))
+                                                    (progn
+                                                      (setq forms (cst:rest forms))
+                                                      (parse-inline form source))
+                                                    (parse-error "Malformed inline form"
+                                                                 (note source form "expected list")))))
+                           :for method := (parse-instance-method-definition (cst:first forms) (cst:second form) source)
+                           :collect
+                           (progn
+                             (when inline-p
+                               (setf (instance-method-definition-inline-p method) inline-p))
+                             (setq forms (cst:rest forms))
+                             method))))
+
```

Comment:
```

:for inline-p := (and (should-inline-p forms) 
                     (prog1 (parse-inline (cst:first forms) source)
                         (setf forms (cst:rest forms))))
:for method := (parse-instance-method-definition (cst:first forms) (cst:second form) source)
:when inline-p :do (setf (instance-method-definition-inline-p method) inline-p)
:collect :method 
:do (setf forms (cst:rest forms))

```

nit: this seems easier to follow to me. managing the state of `forms` is a little confusing to follow, I think something like the above makes it clearer.

Reply:
> nit: this seems easier to follow to me. managing the state of forms is a little confusing to follow, I think something like the above makes it clearer.

Unless I'm missing something, the suggested code still manually manipulates the state of `forms`. But I understand, this part of the code is a bit complicated to follow.

Reply:
that's correct. It's just a nit. Changes are not necessary